net_transport.go

```diff
@@ -568,10 +568,18 @@ func (n *NetworkTransport) handleCommand(r *bufio.Reader, dec *codec.Decoder, en
 		rpc.Command = &req
 
 		// Check if this is a heartbeat
-		if req.Term != 0 && req.Leader != nil &&
-			req.PrevLogEntry == 0 && req.PrevLogTerm == 0 &&
-			len(req.Entries) == 0 && req.LeaderCommitIndex == 0 {
-			isHeartbeat = true
+		if req.ProtocolVersion > 3 {
+			if req.Term != 0 && req.Addr != nil &&
+				req.PrevLogEntry == 0 && req.PrevLogTerm == 0 &&
+				len(req.Entries) == 0 && req.LeaderCommitIndex == 0 {
+				isHeartbeat = true
+			}
+		} else {
+			if req.Term != 0 && req.Leader != nil &&
```

Comment:
Do we still populate the `req.Leader` field? I'm wondering if it would be safe to continue to trust the `req.Leader` field in these cases, instead of having a lot of branches that check the protocol version.

There are a few places where we have this check, but it seems like we could continue to use `req.Leader` for most of these. Is there a reason not to do that?

Either way I think we should add to the `Leader` field godoc, to indicate it is ignored in V4+, or that we continue to use it instead of the `RPCHeader.Addr` (if that is an acceptable option).

Reply:
My understanding is that we would like to remove this code at some point, I think the premise is that we keep 2 versions compatibility in the code. @mkeeler can give more details on this.
I will add the godoc for it.